File: js/managers/resultManager.js
const fs = require('fs');
const path = require('path');

class ResultManager {
    /**
     * Handles saving scraping results to CSV and JSON files.
     */
    constructor(csvPath = 'job_results.csv', jsonPath = null) {
        this.csvPath = csvPath;
        this.jsonPath = jsonPath || path.join(path.dirname(csvPath), path.basename(csvPath, path.extname(csvPath)) + '.json');
        fs.mkdirSync(path.dirname(this.csvPath), { recursive: true });
    }

    saveToCSV(results, headers = null) {
        /**
         * Save results to a CSV file.
         */
        headers = headers || [
            "Job Title", "Company", "Location", "Footer",
            "Easy Apply", "Job Type", "Description",
            "Experience Required", "Salary Mentioned", "Apply Link"
        ];
        const rows = [headers, ...results];
        const csvContent = rows.map(row => row.map(item => `"${String(item).replace(/"/g, '""')}"`).join(',')).join('\n');
        fs.writeFileSync(this.csvPath, csvContent, { encoding: 'utf-8' });
        console.log(`✅ Saved results to CSV: ${this.csvPath}`);
    }

    saveToJSON(results) {
        /**
         * Save results to a JSON file.
         */
        fs.writeFileSync(this.jsonPath, JSON.stringify(results, null, 2), { encoding: 'utf-8' });
        console.log(`✅ Saved results to JSON: ${this.jsonPath}`);
    }
}

module.exports = ResultManager;

File: python/managers/result_manager.py
import csv
import os
import json

class ResultManager:
    """
    Handles saving scraping results to CSV and JSON files.
    """

    def __init__(self, csv_path="job_results.csv", json_path=None):
        self.csv_path = csv_path
        self.json_path = json_path or os.path.splitext(csv_path)[0] + ".json"
        os.makedirs(os.path.dirname(self.csv_path) or ".", exist_ok=True)

    def save_to_csv(self, results, headers=None):
        """
        Save results to a CSV file.
        """
        headers = headers or [
            "Job Title", "Company", "Location", "Footer",
            "Easy Apply", "Job Type", "Description",
            "Experience Required", "Salary Mentioned", "Apply Link",
            "Job ID", "Job URL", "Company-Title Hash"
        ]
        with open(self.csv_path, "w", newline="", encoding="utf-8") as f:
            writer = csv.writer(f)
            writer.writerow(headers)
            writer.writerows(results)
        print(f"✅ Saved results to CSV: {self.csv_path}")

    def save_to_json(self, results):
        """
        Save results to a JSON file.
        """
        with open(self.json_path, "w", encoding="utf-8") as f:
            json.dump(results, f, indent=2, ensure_ascii=False)
        print(f"✅ Saved results to JSON: {self.json_path}")
